misc/util.py

```diff
@@ -84,78 +88,108 @@ def xrun(cmd, *args, **kwargs):
+        logger.debug(
+            "cmd success: %s\n%s\n%s\n",
+            cmdr,
+            maybedecode(stdout_data),
+            maybedecode(stderr_data),
+        )
+
 
 def atexitrun(cmd, *args, **kwargs):
-    cargs = [cmd]+list(args)
+    cargs = [cmd] + list(args)
     atexit.register(xrun, *cargs, **kwargs)
 
+
 def find_indexer(indexer_bin, exc=True):
     if indexer_bin:
         return indexer_bin
     # manually search local build and PATH for algorand-indexer
-    path = ['cmd/algorand-indexer'] + os.getenv('PATH').split(':')
+    path = ["cmd/algorand-indexer"] + os.getenv("PATH").split(":")
     for pd in path:
-        ib = os.path.join(pd, 'algorand-indexer')
+        ib = os.path.join(pd, "algorand-indexer")
         if os.path.exists(ib):
             return ib
-    msg = 'could not find algorand-indexer. use --indexer-bin or PATH environment variable.'
+    msg = "could not find algorand-indexer. use --indexer-bin or PATH environment variable."
     if exc:
         raise Exception(msg)
     logger.error(msg)
     return None
 
+
 def ensure_test_db(connection_string, keep_temps=False):
     if connection_string:
         # use the passed db
         return connection_string
     # create a temporary database
-    dbname = 'e2eindex_{}_{}'.format(int(time.time()), random.randrange(1000))
-    xrun(['dropdb', '--if-exists', dbname], timeout=5)
-    xrun(['createdb', dbname], timeout=5)
+    dbname = "e2eindex_{}_{}".format(int(time.time()), random.randrange(1000))
+    xrun(["dropdb", "--if-exists", dbname], timeout=5)
+    xrun(["createdb", dbname], timeout=5)
     if not keep_temps:
-        atexitrun(['dropdb', '--if-exists', dbname], timeout=5)
+        atexitrun(["dropdb", "--if-exists", dbname], timeout=5)
     else:
         logger.info("leaving db %r", dbname)
-    return 'dbname={} sslmode=disable'.format(dbname)
+    return "dbname={} sslmode=disable".format(dbname)
+
 
 # whoever calls this will need to import boto and get the s3 client
-def firstFromS3Prefix(s3, bucket, prefix, desired_filename, outdir=None, outpath=None):
-    response = s3.list_objects_v2(Bucket=bucket, Prefix=prefix, MaxKeys=10)
-    if (not response.get('KeyCount')) or ('Contents' not in response):
-        raise Exception('nothing found in s3://{}/{}'.format(bucket, prefix))
-    for x in response['Contents']:
-        path = x['Key']
-        _, fname = path.rsplit('/', 1)
+def firstFromS3Prefix(
+    s3, bucket, prefix, desired_filename, outdir=None, outpath=None
+) -> bool:
+    haystack = []
+    found_needle = False
+
+    def report():
+        print(
+            f"""Hello from firstFromS3Prefix() !!!!
+        I finished searching for the needle {desired_filename} in the AWS S3 path {bucket}/{prefix}.
+        haystack={haystack}
+        len(haystack)={len(haystack)}
+        found_needle={found_needle}
+"""
+        )
+
+    atexit.register(report)
+
+    response = s3.list_objects_v2(Bucket=bucket, Prefix=prefix, MaxKeys=100)
```

Comment:
If this line was the only actual change, it would be trivial to approve.

Reply:
There's a bit more, but only in that function.